gpu4pyscf/nac/tdrhf.py

```diff
@@ -0,0 +1,302 @@
+        v1ao = vresp(dm + dm.T)
+        return reduce(cp.dot, (orbv.T, v1ao, orbo)).ravel()
+
+    z1 = cphf.solve(
+        fvind,
+        mo_energy,
+        mo_occ,
+        -LI*1.0*EI, # only one spin, negative in cphf
+        max_cycle=td_nac.cphf_max_cycle,
+        tol=td_nac.cphf_conv_tol)[0] # eq.(83) in Ref. [1]
+
+    z1 = z1.reshape(nvir, nocc)
+    z1ao = reduce(cp.dot, (orbv, z1, orbo.T)) * 2 # double occupency
+    # eq.(50) in Ref. [1]
+    z1aoS = (z1ao + z1ao.T)*0.5 # 0.5 is in the definition of z1aoS
+    # eq.(73) in Ref. [1]
+    GZS = vresp(z1aoS) # generate the double occupency 
+    GZS_mo = reduce(cp.dot, (mo_coeff.T, GZS, mo_coeff))
+    W = cp.zeros((nmo, nmo))  # eq.(75) in Ref. [1]
+    W[:nocc, :nocc] = GZS_mo[:nocc, :nocc]
+    zeta0 = mo_energy[nocc:, cp.newaxis]
+    zeta0 = z1 * zeta0
+    W[:nocc, nocc:] = GZS_mo[:nocc, nocc:] + 0.5*yI.T*EI + 0.5*zeta0.T #* eq.(43), (56), (28) in Ref. [1]
+    zeta1 = mo_energy[cp.newaxis, :nocc]
+    zeta1 = z1 * zeta1
+    W[nocc:, :nocc] = 0.5*xI*EI + 0.5*zeta1
+    W = reduce(cp.dot, (mo_coeff, W , mo_coeff.T)) * 2.0
+
+    mf_grad = mf.nuc_grad_method()
+    s1 = mf_grad.get_ovlp(mol)
+    dmz1doo = z1aoS
+    oo0 = reduce(cp.dot, (orbo, orbo.T)) * 2.0
+
+    if atmlst is None:
+        atmlst = range(mol.natm)
+    
+    hcore_deriv = pyscf.grad.rhf.hcore_generator(mf_grad.to_cpu(), mol)
+    offsetdic = mol.offset_nr_by_atom()
+    de = cp.zeros((len(atmlst),3))
+    de_etf = cp.zeros((len(atmlst),3))
+    xIao = reduce(cp.dot, (orbo, xI.T, orbv.T)) * 2
+    yIao = reduce(cp.dot, (orbv, yI, orbo.T)) * 2
+    eri1 = -mol.intor('int2e_ip1', aosym='s1', comp=3)
+    eri1 = eri1.reshape(3,nao,nao,nao,nao)
+    for k, ia in enumerate(atmlst): # eq.(58) in Ref. [1]
+        shl0, shl1, p0, p1 = offsetdic[ia]
+        h1ao = hcore_deriv(ia)
+        h1ao = cp.asarray(h1ao)
+        s1_tmp = s1.copy()
+        s1_tmp[:,:p0] = 0
+        s1_tmp[:,p1:] = 0
+        s1_tmp = s1_tmp + s1_tmp.transpose(0,2,1)
+
+        eri1a = eri1.copy()
+        eri1a[:,:p0] = 0
+        eri1a[:,p1:] = 0
+        eri1a = eri1a + eri1a.transpose(0,2,1,3,4)
+        eri1a = eri1a + eri1a.transpose(0,3,4,1,2)
+        erijk = eri1a - eri1a.transpose(0,1,4,3,2)*0.5
+
+        ds1_tmp = s1.copy()
+        ds1_tmp = ds1_tmp.transpose(0,2,1)
+        ds1_tmp[:,:,:p0] = 0
+        ds1_tmp[:,:,p1:] = 0
+
+        de[k] = cp.einsum('xpq,pq->x', h1ao, dmz1doo)
+        de[k] -= cp.einsum('xpq,pq->x', s1_tmp, W)
+        de[k] += cp.einsum('xijkl,ij,kl->x', erijk, oo0, dmz1doo)
+        de_etf[k] = de[k]
+        de[k] += cp.einsum('xij,ij->x', ds1_tmp, xIao*EI)
+        de[k] += cp.einsum('xij,ij->x', ds1_tmp, yIao*EI)
+        de_etf[k] += cp.einsum('xij,ij->x', s1_tmp, xIao*EI) * 0.5
+        de_etf[k] += cp.einsum('xij,ij->x', s1_tmp, yIao*EI) * 0.5
+    
+    de = de.get()
+    de_etf = de_etf.get()
+    return de, de/EI, de_etf, de_etf/EI
+
+
+class NAC(rhf_grad.GradientsBase):
```

Comment:
An independent class, subclass to lib.StreamObject. no need to inherit to Gradients. If any methods of the base Gradients class are needed, assign them within this new class, e.g.
```
get_ovlp = GradientBase.get_ovlp
```